[Woo POS] Add test coverage for PointOfSaleDashboardViewModel

soundsokay opened the pull request:
Expands test coverage for PointOfSaleDashboardViewModel to verify the `observePaymentStateForButtonDisabledProperties` method correctly sets `isAddMoreDisabled` accounting for both payment state and order syncing state.

<!-- Remember about a good descriptive title. -->

Closes: #13306

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
The Add More button was disabled in #13270 but tests weren't updated. This PR adds coverage for that.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
Invert the values returned and used in the assignment to `isAddMoreDisabled`, run the `PointOfSaleDashboardViewModelTests`, and observe failures in the newly added tests.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

soundsokay committed:
test: Add test coverage for PointOfSaleDashboardViewModel

Verify isAddMoreDisabled is set appropriately for payment state and order syncing state.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

bozidarsevo commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -36,13 +36,12 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
          orderService: POSOrderServiceProtocol,
-         currencyFormatter: CurrencyFormatter) {
+         currencyFormatter: CurrencyFormatter,
+         totalsViewModel: TotalsViewModel) {
```
Since with this change the totalsViewModel is injected and not created, what was the main reason for it? And will we need to change some other parts of the POS code with this?

soundsokay replied:
It was a first-cut attempt to improve testability of this class.

This injection was removed in commit 35370c72d7b0f5f967bfaa9304c256416372c12f

iamgabrielma commented on WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift:
```diff
@@ -24,9 +24,12 @@ final class TotalsViewModel: ObservableObject {
-    @Published private(set) var isSyncingOrder: Bool = false
 
-    @Published private(set) var paymentState: PaymentState = .acceptingCard
+    /// This is not private so that PointOfSaleDashboardViewModelTests will compile
+    @Published var isSyncingOrder: Bool = false
+
+    /// This is not private so that PointOfSaleDashboardViewModelTests will compile
+    @Published var paymentState: PaymentState = .acceptingCard
```
I think we should't modify the access control here for the sake of testing, if these do not allow for property injection when the app's logic runs, they shouldn't be allowed either for the tests. By making them internal just for testing, now we're changing the implementation and these become settable in the app's target. Eg:

```
totalsViewModel.isSyncingOrder = isSyncingOrder
```

We wouldn't want that, as now anyone can change the state on demand. Ideally we'd want the view model to handle its own state and be in `isSyncingOrder` state as a consequence of other actions, and keep it inaccessible otherwise, then we assert the state by triggering actions that would change this.

staskus replied:
> I think we should't modify the access control here for the sake of testing, if these do not allow for property injection when the app's logic runs, they shouldn't be allowed either for the tests

Agree. `paymentState` changes, when `cardPresentPaymentService.paymentEventPublisher` emits an event. `isSyncingOrder` changes during different stages of `syncOrder` flow. This is what we want to test.

soundsokay replied:
Good point @iamgabrielma. The original access controls were restore in commit 7fc1fc340a0d4699b7a0e50826902ebd40c56d9b

Thanks for your feedback @staskus. How is commit 72f6822349b505fafa6ddd8ddbbd5e7e5abfefcf looking as a better approach for this test coverage expansion?

staskus replied:
Thanks.

Unfortunately, this is not the right approach. We shouldn't create simulation methods within the protocol for unit testing. We should be calling **real** methods when testing. 

For example, when we unit test `PointOfSaleDashboardViewModel`, our `sut` has to be a real `PointOfSaleDashboardViewModel` but its dependencies should be mocked. Mocking also happens within the testing target and not within the main app target or any of the other frameworks.

In this case, we want to test how the `PointOfSaleDashboardViewModel` behaves:

Order is syncing when `cartViewModel cartSubmissionPublisher` emits an event.

```
    func observeCartSubmission() {
        cartViewModel.cartSubmissionPublisher.sink { [weak self] cartItems in
            guard let self else { return }
            orderStage = .finalizing
            startSyncingOrder(cartItems: cartItems)
        }
        .store(in: &cancellables)
    }
```

For testing, we should inject mocked dependencies into  `PointOfSaleDashboardViewModel`:
- Inject a mocked `CartViewModel` and then we manually emit an event through `cartSubmissionPublisher`
- Inject a mocked `TotalsViewModel` to verify that when `cartSubmissionPublisher` emits an event, `startSyncingOrder` is called, and in other tests emit events through `paymentState`, or `isSyncingOrder` to test how `PointOfSaleDashboardViewModel` state changes. 

To mock `CartViewModel` or `TotalsViewModel` you can use `protocols`. `PointOfSaleDashboardViewModel` will only know about abstractions, so you can mock them in the testing target.

soundsokay replied:
Thanks for your feedback. I pushed up another set of changes and took care not to modify the class under test.

The  tests now mock dependencies that are injected.

staskus replied:
Thank you! 🙇 I will check the changes!

iamgabrielma commented on WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift:
```diff
@@ -73,7 +76,9 @@ final class TotalsViewModel: ObservableObject {
          cardPresentPaymentService: CardPresentPaymentFacade,
-         currencyFormatter: CurrencyFormatter) {
+         currencyFormatter: CurrencyFormatter,
+         paymentState: PaymentState = .acceptingCard,
+         isSyncingOrder: Bool = false) {
```
By setting these two properties and its initial values on initialization we're now setting some default initial state, when before that it was stateless. Note how the only dependencies being injected are abstractions like the payments Façade, utils like the currency formatter, or the POS interfaces.

Do we want it to always default to `.acceptingCard` and not `syncing`? Would it be preferable to leave the implementation details to each service? Eg: Payment state being set and updated from the `CardPresentPaymentService` rather than injected in.

soundsokay replied:
Looks like these have [initial values on trunk](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift#L27)?

iamgabrielma replied:
> Looks like these have [initial values on trunk](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift#L27)?

They are, but it's also internal state to the class. A concern here would be: 

Imagine we implement a new XYZ that needs the `TotalsViewModel` to be in `.acceptingCard` state. Since the property is exposed, and this happens to be its default value then we just ignore setting the property specifically. Later on we update the implementation details of the class to be `.notAcceptingCard` instead, which will cause XYZ to fail because it relied on default state that is no longer the default case, and would trigger no errors (unless catch by unit tests!) because the compiler would find no issues with it.

To avoid these, we can either force providing a specific initial state by not providing a default value, so the compiler enforces this, or in our case just not expose the state to be modified from outside the class, and leave the responsibility of it to the `TotalsViewModel` itself, opening some getters so it can be checked (but not modified) for other objects that need to check it.

soundsokay replied:
That's a a good point.

Addressed in commit bb3d973460da54ed4d807a56d91978a735e88840.

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -126,10 +126,14 @@ import Combine
     // https://github.com/woocommerce/woocommerce-ios/issues/13207
     let orderStageSubject = PassthroughSubject<PointOfSaleDashboardViewModel.OrderStage, Never>()
     let orderStagePublisher = orderStageSubject.eraseToAnyPublisher()
+    let totalsViewModel = TotalsViewModel(orderService: POSOrderPreviewService(),
+                                              cardPresentPaymentService: CardPresentPaymentPreviewService(),
+                                              currencyFormatter: .init(currencySettings: .init()))
```
While this happens on the `#Preview`, declaring a new instance of `TotalsViewModel` in order to inject it into the VM for the `CartView` seems to indicate that the design of the dependency graph is not quite right. I'd say the `CartView` should know nothing about the Totals VM and perhaps this shouldn't be injected here.

soundsokay replied:
Addressed in commit 35370c72d7b0f5f967bfaa9304c256416372c12f.

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -124,11 +124,15 @@ fileprivate extension CardPresentPaymentEvent {
 #if DEBUG
 #Preview {
     NavigationStack {
+        let totalsViewModel = TotalsViewModel(orderService: POSOrderPreviewService(),
+                                                      cardPresentPaymentService: CardPresentPaymentPreviewService(),
+                                                      currencyFormatter: .init(currencySettings: .init()))
```
Same as above, up until now the entry point seems to accept only abstractions either from the payment façade, or the POS interfaces, however now it would require to inject a specific POS VM.

soundsokay replied:
Addressed in commit 35370c72d7b0f5f967bfaa9304c256416372c12f.

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift:
```diff
@@ -13,26 +14,96 @@ final class PointOfSaleDashboardViewModelTests: XCTestCase {
         super.tearDown()
     }
 
+    private func setupViewModelWithExpectationsExpectingAddMoreMatchesSyncingState(paymentState: TotalsViewModel.PaymentState,
```
Being helper methods, these could go into an extension. I think there's also no need in the test target for these to be private. Maybe we can also simplify the function names? Since it would seem are specific only to the `AddMoreIsDisabled` property, however both payment state and syncing state are exposed.  Eg:
```
setupViewModel(paymentState: TotalsViewModel.PaymentState, isSyncingOrder: Bool)
```

iamgabrielma reviewed (commented):
Thanks for working on this! I left a couple of comments regarding the decision of injecting the `TotalsViewModel`, as well as exposing private properties that were only gettable. 

Perhaps we can improve the test coverage here by using Mocks/Fakes/Stubs/Spies instead? As an example we do this with the [item provider](https://github.com/woocommerce/woocommerce-ios/blob/74454c0ec4e8b91a3578590888faa73acfe18852/WooCommerce/WooCommerceTests/POS/ViewModels/ItemListViewModelTests.swift#L188-L202)

soundsokay committed:
test: Remove TotalsViewModel dependency injection

soundsokay committed:
test: Restore prior access controls

Keep isSyncingOrder and paymentState private.

soundsokay committed:
test: simulate order syncing to verify Add More button is disabled when payment state is idle

Added PointOfSaleDashboardViewModelProtocol to simulate order syncing for use in the unit test.

soundsokay committed:
chore: move protocol to a separate file

soundsokay committed:
test: improve expectation description

soundsokay committed:
test: verify the Add More button is disabled after collect payment success

soundsokay committed:
test: improve test name with camelCase

staskus commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -159,3 +159,10 @@ private extension PointOfSaleDashboardViewModel {
     }
 }
+
+// Conform to the protocol needed for unit testing
```
We should be using protocols for real methods, not for testing private methods. We should be likely covering `PointOfSaleDashboardViewModel` dependencies with protocols, not the view model itself. More explanation here: https://github.com/woocommerce/woocommerce-ios/pull/13310#discussion_r1680493636

staskus commented on WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift:
```diff
@@ -81,6 +80,46 @@ final class PointOfSaleDashboardViewModelTests: XCTestCase {
+    func test_isAddMoreDisabled_is_true_when_order_is_syncing_and_payment_state_is_idle() {
+        var sut: any PointOfSaleDashboardViewModelProtocol
+
+        sut = PointOfSaleDashboardViewModel(itemProvider: itemProvider,
```
As explained in https://github.com/woocommerce/woocommerce-ios/pull/13310#discussion_r1680493636, we should just use `PointOfSaleDashboardViewModel` within `PointOfSaleDashboardViewModelTests`. However, all the dependencies that `PointOfSaleDashboardViewModel` depend on, can be mocked and used here.

staskus commented on WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift:
```diff
@@ -81,6 +80,46 @@ final class PointOfSaleDashboardViewModelTests: XCTestCase {
         XCTAssertEqual(sut.isCartCollapsed, expectedCartCollapsedState)
     }
 
+    func test_isAddMoreDisabled_is_true_when_order_is_syncing_and_payment_state_is_idle() {
```
If we look at what `PointOfSaleDashboardViewModel` does, it actually doesn't change `isAddMoreDisabled` state when the order is syncing.

It changes it when `totalsViewModel` changes `paymentState` and `isSyncingOrder`. When we're unit testing `PointOfSaleDashboardViewModel`, we're oblivious at how `totalsViewModel` behaves internally. 

When we write tests for `totalsViewModel`, then we can assert that when `totalsViewModel.startSyncingOrder`, then `paymentState` and/or `isSyncingOrder` changes.

soundsokay committed:
test: refactor to make PointOfSaleDashboardViewModel testable

Work in progress.

soundsokay committed:
chore: remove TotalsViewModelProtocol from test target

Fixes a compile issue.

soundsokay committed:
test: simulate order syncing using mocked dependencies

Fixes test: test_isAddMoreDisabled_is_true_when_order_is_syncing_and_payment_state_is_idle

soundsokay committed:
test: expand test coverage for PointOfSaleDashboardViewModel

Adds: test_isAddMoreDisabled_is_true_for_processingPayment.

soundsokay committed:
chore: remove PointOfSaleDashboardViewModelProtocol

Protocol no longer needed after refactoring to new testing approach using mocked dependencies.

soundsokay committed:
test: inject a CartViewModel? into PointOfSaleDashboardViewModel

Also restore read-only access control for external entities with lazy initialization for cartViewModel.

soundsokay committed:
chore: remove comments

soundsokay committed:
test: expand test coverage using mocked dependencies

staskus commented on WooCommerce/Classes/POS/ViewModels/AnyTotalsViewModel.swift:
```diff
@@ -0,0 +1,138 @@
+import Combine
+import Yosemite
+
+class AnyTotalsViewModel: ObservableObject, TotalsViewModelProtocol {
```
Mmm... I forgot about the complication of using `ObservableObject` as a protocol. Sorry for suggesting it all before thinking more deeply 🤔 

I'm concerned about initializers that assign publishers. It could be easy to forget to do and it requires duplication any time we add something new.

However, the only other solution that I know of is using generics which also adds a bit of overhead. 

I tried it because I didn't want to suggest anything huge without knowing that it would work:

https://github.com/woocommerce/woocommerce-ios/compare/feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test...feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test-test?expand=1

Let me know what are you thoughts about this wrapper/generics usage/some other path.

 Generics work but it also shows how `TotalsViewModel`, `CardViewModel`, and `PointOfSaleDashboardViewModel` are all referencing each other. It may make sense to review the dependency graph at the end of M2 to make sure it's all necessary and no retain cycles are created.

soundsokay replied:
Thanks for your review. 

> it also shows how TotalsViewModel, CardViewModel, and PointOfSaleDashboardViewModel are all referencing each other

> I forgot about the complication of using ObservableObject as a protocol. 

Indeed. Since taking on this work I've struggled with tradeoffs.

My current set of changes which properly mock dependencies produce a larger diff and have other downstream effects as you've noted than my earlier approach to adding test coverage which were a pragmatic solution but required undesirable changes to the class under test. 

In the latest approach `PointOfSaleDashboardViewModel` is in better shape, but some complexity cropped up while abstracting its dependencies.

Regarding

> shows how TotalsViewModel, CardViewModel, and PointOfSaleDashboardViewModel are all referencing each other.

This goes to the heart of the matter. It look like `PointOfSaleDashboardViewModel` wasn't designed with testing in mind. It had tight coupling between `PointOfSaleDashboardViewModel`, `CartViewModel`, and `TotalsViewModel` which I've reduced by introducing abstractions including interface-based interactions, type erasure and dependency injection.

>review the dependency graph at the end of M2 to make sure it's all necessary

I too wonder if a more fundamental refactoring is in order at some point.

In the meantime though I'll take a look at your suggestion of using generics. Thanks for that.

staskus replied:
Yes, I understand the complexities now after touching the code. The setup is hard to test which shows problems with decoupling. We tried to decouple the logic by splitting it into different classes or files but it all still remains coupled and just harder to test or use.

> In the meantime though I'll take a look at your suggestion of using generics. Thanks for that.

Okay, unfortunately, it also introduces a bunch of cascading changes due to same reasons. However, if we continue using observable VMs within other VMs, then generics may be the only way that allows us to inject mocks cleanly.

soundsokay replied:
Thanks again for your thorough review and draft solution using generics. 

I agree that using generics could help inject mocks cleanly and address the issue of assigning publishers in initializers. However, this approach also introduces a layer of complexity and, as you noted, a different set of cascading changes, which might slow down our progress in the short term.

After evaluating the pros and cons I think my current approach, with improvements to address your concerns about manual publisher initialization in `AnyTotalsViewModel`, is preferable to using generics to accomplish the decoupling needed to expand test coverage.


Here's my rationale.

1. Improved publisher initialization
Commit b14e014047698dbdfaf7df70543b62514d9124f3 responds to your feedback about manual publisher initialization. 

2. Speed of development: Changes from `trunk` have been merged in. Types are more loosely coupled and test coverage of `PointOfSaleDashboardViewModel` expanded using protocols. These make the current implementation a move in the right direction.

3. Improved readability and lower overall complexity: I find the protocol-based code more readable and easier to understand. It doesn't have the syntax verbosity that comes with generics. 
4. Precedence in the codebase. It looks like there are limited example in the codebase of [using generics for view models](https://github.com/woocommerce/woocommerce-ios/blob/10900a66bb13b3c611590d66acae6cbcc4c72858/WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift#L23). Just one I think based on a search of `ViewModel<`. However searching `ViewModel(` yields many results.

Regarding:
>However, if we continue using observable VMs within other VMs, then generics may be the only way that allows us to inject mocks cleanly.

This isn't a pattern I would recommend we continue to follow. Being the exception rather than the rule can we reserve generics for this situation if it continues to arise?

staskus replied:
Thanks. Yes, I agree with you that in the context of `generics` vs `type-erased AnyVM`, the latter has sound arguments. 

However, by looking at the solution again, it still feels like we're trying to go against the system frameworks and the language. I want to avoid us choosing an overcomplicated solution and slowing down our M2 work.

We want to keep things testable but also we want keep things simple.

---

I'll try to go through the problem again.

#### What is the problem?

We want to make our ViewModel code testable.

#### What is the common solution?

Use protocols and dependency injection.

#### What stopping us from using it?

Using protocols and `@ObservedObject` is not supported.

#### When it happens?

For example, in `PointOfSaleDashboardView` we're referencing `TotalsViewModel`. But we also want to inject `TotalsViewModel` into `PointOfSaleDashboardViewModel` for testability.

And I think that's the main problem. `SwiftUI` expects `View Model` to be used by `View`. However, we want to use `ViewModel` within `Views` AND within other `ViewModels`.

#### What can we do about it?

Let's use a concrete implementation in `Views`, and protocols in `View Models`. 

It's a short-term move that doesn't force us into major refactorings, using generics, or opting for `AnyViewModel` type of solutions.

The side effect is that it forces us to clean up dependency management and to make sure all of the view models are created in a single place. It transparently exposes the fact that our code is not decoupled but nothing will stop us from refactoring in the future.

Here's a WIP:

https://github.com/woocommerce/woocommerce-ios/compare/feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test...feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test-rearchitect?expand=1

I didn't update tests but I think it would work. The main changes are:

- Creating dependencies at entry point
- Passing dependencies directly to views and view models

It allows to expect protocols in VMs and concrete implementation in Views.

Let me know what you think.

soundsokay committed:
test: refactor AnyTotalsViewModel to centralize publisher bindings

- Converted publishers to computed properties leveraging property wrappers for simplified exposure and automatic initialization.
- Introduced a `bindPublishers` method to centralize the assignment of publishers, ensuring all necessary bindings are in one place.
- Addressed feedback by eliminating manual publisher initialization in initializers, reducing duplication and potential errors.
- Enhanced code maintainability and readability by using property wrappers and centralizing publisher assignments.

soundsokay committed:
Merge remote-tracking branch 'origin/trunk' into feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test

Conflicts:
	WooCommerce/Classes/POS/Presentation/TotalsView.swift
	WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift

soundsokay committed:
refactor: ensure TotalsViewModel initializes without default values

- Updated the TotalsViewModel to remove default values for `paymentState` and `isSyncingOrder`, making these parameters mandatory during initialization.

bozidarsevo commented on WooCommerce/Classes/POS/ViewModels/AnyTotalsViewModel.swift:
```diff
@@ -0,0 +1,155 @@
+    var formattedOrderTotalPricePublisher: Published<String?>.Publisher { $_formattedOrderTotalPrice }
+    var formattedOrderTotalTaxPricePublisher: Published<String?>.Publisher { $_formattedOrderTotalTaxPrice }
+
+    private var wrapped: any TotalsViewModelProtocol
```
If there is a `TotalsViewModelProtocol` protocol, can we remove this `AnyTotalsViewModel`? Since this `AnyTotalsViewModel ` conforms to `TotalsViewModelProtocol` and has a property of type `TotalsViewModelProtocol`... that feels a bit weird..
A lot of code here in `AnyTotalsViewModel` is just calls to `wrapped.something...`

staskus replied:
We discussed it here https://github.com/woocommerce/woocommerce-ios/pull/13310#discussion_r1684167411

staskus replied:
I agree it's so far from ideal. It makes it incredibly hard for me to proceed with the solution. Any work we want to do in TotalsViewModel will now have to duplicate within protocol AND within AnyTotalsViewModel. 🤔 Any more ideas or thoughts from your side?

It likely indicates that our architectural approach of reusing the view models throughout different views and also reusing the same view models within other view models goes "against the stream" of SwiftUI development.

soundsokay replied:
Thanks for your comments @bozidarsevo and @staskus 

Clearly making the `PointOfSaleDashboardViewModel` testable is putting pressure on the existing code design. I might need to look at refactoring to take care of that underlying issue, unless we want to try the solution using generics for dependency injection, which as noted above has it's own drawbacks.

staskus replied:
@soundsokay I made another suggestion. I think we can avoid both generics, AnyViewModel, and a lengthy refactoring.

bozidarsevo reviewed (commented):
I am wondering if we need `AnyTotalsViewModel` and can just go with `TotalsViewModelProtocol` and having `TotalsViewModel` and `MockTotalsViewModel` conform to it...

staskus committed:
Use protocols in ViewModels and concrete implementations in Views

staskus reviewed (commented):
I made another suggestion.

I hate the fact that I'm holding this PR for so long but I want to make sure we don't make things harder for ourselves. 

I suggest to use concrete implementation in Views, and protocols in View Models. This way we go around any compiler issues that we had and it also transparently exposes our dependency tree that we can refactor in the future.

soundsokay commented:
Thanks for your detailed feedback and WIP solution @staskus.

I too would like to see this moving faster but also want to get this right and not add undesirable patterns while staying pragmatic and mindful of speed of development in the short and long term.

I'll take a look and see if your WIP solution looks like a lower level of effort than the refactoring I had planned.

soundsokay committed:
test: fix compilation issue

- Make tests compile.
- Required changing access modifiers on `PointOfSaleDashboardViewModel` from `private` to internal access.

soundsokay committed:
test: expanded and refactored PointOfSaleDashboardViewModelTests

- Added test_observeCartAddMoreAction_updates_orderStage_to_building.
- Added two tests that were formerly combined and now have been broken out into separate cases.
- Cleaned up other tests.

soundsokay committed:
chore: fix linter issue

Remove trailing whitespace.

soundsokay commented:
Your rearchitecture looked good to me. 

I merged it in and make adjustments to get the tests to compile in commit e8df6378d4ae58f37b34af4dcd54af3d7ab13c04.

One downside is that changed access modifiers for `cartViewModel` and `totalsViewModel` on `PointOfSaleDashboardViewModel` from `private` to internal access. However the class already uses internal access for `itemListViewModel`.

I also expanded test coverage and cleaned up `PointOfSaleDashboardViewModelTests`. Which were my focus before pushing up the changes.

CartViewModelTests are failing. Maybe you can have a look?

Seems the issue is with `CartViewModel` [initializing without any parameters](https://github.com/woocommerce/woocommerce-ios/pull/13310/commits/19bf5a648137f5c1540f93fc0f9028c3f1046d9c#diff-d3b44da1ee099063293649a299b147271137d1c009790b5914d831e1faa0b6deR23).

staskus commented:
> CartViewModelTests are failing. Maybe you can have a look? Seems the issue is with CartViewModel [initializing without any parameters](https://github.com/woocommerce/woocommerce-ios/pull/13310/commits/19bf5a648137f5c1540f93fc0f9028c3f1046d9c#diff-d3b44da1ee099063293649a299b147271137d1c009790b5914d831e1faa0b6deR23).

@soundsokay yes, it could be! `CartViewModel` depended on `POSDashboardVM` passing an observer through the initializer which prevented `CartViewModel` from being initialized separately. I will take a look 👍

staskus committed:
Introduce CartViewModelProtocol to properly mock CartViewModel in tests

staskus committed:
Bind mock orderStageSubect with CartViewModel to fix failing tests

staskus reviewed (approved):
It builds, I think it's good time to merge this 👍 

@bozidarsevo, would be good if you could take one more look yourself, since I also made the last commits.

bozidarsevo committed:
Merge branch 'trunk' into feat/13127-disable-cta-to-go-back-to-product-selector-when-order-is-being-synced-add-test

soundsokay commented:
Thank you for your contributions @staskus and @bozidarsevo. 

I'm taking a look now.

bozidarsevo commented:
I will fix the build now, might have made a wrong conflict merge commit

bozidarsevo committed:
Update CartViewModel.swift

soundsokay committed:
test: expand test coverage

Adds verification that orderStage is set to .building when cart becomes empty.

soundsokay committed:
chore: remove TODO

Remove TODO for adding test coverage for PointOfSaleDashboardViewModel now that it has been added.

soundsokay committed:
chore: fix linting issues

soundsokay commented:
I made recent commits that only change test code in PointOfSaleDashboardViewModelTests to expand test coverage and remove comments and whitespace. 

With all checks passing and two 👍 I'm going to merge.

Thanks again for your reviews and contributions @staskus and @bozidarsevo.
